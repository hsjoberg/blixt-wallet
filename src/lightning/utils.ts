import deepMapKeys from "deep-map-keys";

export const fixGrpcJsonResponse = <T>(response: { [keys: string]: any })  =>
  deepMapKeys<T>(response, (key) => {
    if (key[key.length - 1] === "_") {
      return key.slice(0, -1);
    }
    return "$" + key;
});
