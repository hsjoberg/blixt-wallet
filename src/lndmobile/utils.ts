import { NativeModules } from "react-native";
import deepMapKeys from "deep-map-keys";
import * as base64 from "base64-js";
import * as $protobuf from "protobufjs";
import { lnrpc } from "../../proto/proto";

const { LndMobile } = NativeModules;

export interface ISendRequestClass<IReq, Req> {
  create: (options: IReq) => Req;
  encode: (request: Req) => $protobuf.Writer;
}

export interface ISendResponseClass<Res> {
  decode: (reader: $protobuf.Reader | Uint8Array) => Res;
  toObject(message: Res, options?: $protobuf.IConversionOptions): { [k: string]: any };
}

export interface ISyncCommandOptions<IReq, Req, Res> {
    request: ISendRequestClass<IReq, Req>;
    response: ISendResponseClass<Res>;
    method: string;
    options: IReq;
}

export interface IStreamCommandOptions<IReq, Req> {
    request: ISendRequestClass<IReq, Req>;
    method: string;
    options: IReq;
}

export interface IStreamResultOptions<Res> {
  base64Result: string;
  response: ISendResponseClass<Res>;
}

/**
 * @throws
 */
export const sendCommand = async <IReq, Req, Res>({ request, response, method, options }: ISyncCommandOptions<IReq, Req, Res>): Promise<Res> => {
  try {
    const instance = request.create(options);
    const b64 = await LndMobile.sendCommand(method, base64.fromByteArray(request.encode(instance).finish()));
    return response.decode(base64.toByteArray(b64.data));
  } catch (e) { throw e; }
};

export const sendStreamCommand = async <IReq, Req>({ request, method, options }: IStreamCommandOptions<IReq, Req>, streamOnlyOnce: boolean = false): Promise<string> =>  {
  const instance = request.create(options);
  const response = await LndMobile.sendStreamCommand(method, base64.fromByteArray(request.encode(instance).finish()), streamOnlyOnce);
  return response;
};

export const decodeStreamResult = <Res>( { base64Result, response }: IStreamResultOptions<Res>): Res => {
  return response.decode(base64.toByteArray(base64Result));
};
