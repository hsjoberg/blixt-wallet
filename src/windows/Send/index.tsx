import React from "react";
import { createStackNavigator } from "@react-navigation/stack";
import SendCamera from "./SendCamera";
import SendConfirmation from "./SendConfirmation";

const Stack = createStackNavigator();

export type SendStackParamList = {
  SendCamera: undefined;
  SendConfirmation: undefined;
}

export default () => {
  return (
    <Stack.Navigator initialRouteName="SendCamera" screenOptions={{ headerShown: false, animationEnabled: false }}>
      <Stack.Screen name="SendCamera" component={SendCamera} />
      <Stack.Screen name="SendConfirmation" component={SendConfirmation} />
    </Stack.Navigator>
  )
}
