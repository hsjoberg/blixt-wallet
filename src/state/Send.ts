import * as Bech32 from "bech32";

import { Action, Thunk, action, thunk } from "easy-peasy";
import { getGeolocation, hexToUint8Array } from "../utils";
import { lnrpc, routerrpc } from "../../proto/lightning";

import { ILNUrlPayResponse } from "./LNURL";
import { IStoreInjections } from "./store";
import { IStoreModel } from "./index";
import { ITransaction } from "../storage/database/transaction";
import { LnBech32Prefix } from "../utils/build";
import Long from "long";
import { PLATFORM } from "../utils/constants";
import { identifyService } from "../utils/lightning-services";
import logger from "./../utils/log";
import { valueFiat } from "../utils/bitcoin-units";

let ReactNativePermissions: any;
if (PLATFORM !== "macos") {
  ReactNativePermissions = require("react-native-permissions");
}

const log = logger("Send");

type PaymentRequest = string;

export interface ISendModelSetPaymentPayload {
  paymentRequestStr: PaymentRequest;
  extraData?: IExtraData;
}

export interface IModelSendPaymentPayload {
  amount?: Long;
  outgoingChannelId?: Long;
}

export interface IModelQueryRoutesPayload {
  amount: Long;
  pubKey: string;
  routeHints?: lnrpc.IRouteHint[];
}

interface IExtraData {
  payer: string | null;
  type: ITransaction["type"];
  website: string | null;
  lnurlPayResponse: ILNUrlPayResponse | null;
  lightningAddress: string | null;
  lud16IdentifierMimeType: string | null;
  lnurlPayTextPlain: string | null;
}

export interface ISendModel {
  clear: Action<ISendModel>;
  setPayment: Thunk<
    ISendModel,
    ISendModelSetPaymentPayload,
    IStoreInjections,
    {},
    Promise<lnrpc.PayReq>
  >;
  sendPayment: Thunk<
    ISendModel,
    IModelSendPaymentPayload | void,
    IStoreInjections,
    IStoreModel,
    Promise<lnrpc.Payment>
  >;
  queryRoutesForFeeEstimate: Thunk<
    ISendModel,
    IModelQueryRoutesPayload,
    IStoreInjections,
    IStoreModel,
    Promise<lnrpc.QueryRoutesResponse>
  >;

  setPaymentRequestStr: Action<ISendModel, PaymentRequest>;
  setPaymentRequest: Action<ISendModel, lnrpc.PayReq>;
  setRemoteNodeInfo: Action<ISendModel, lnrpc.NodeInfo>;
  setExtraData: Action<ISendModel, IExtraData>;

  paymentRequestStr?: PaymentRequest;
  remoteNodeInfo?: lnrpc.NodeInfo;
  paymentRequest?: lnrpc.PayReq;
  extraData?: IExtraData;
}

export const send: ISendModel = {
  clear: action((state) => {
    state.paymentRequestStr = undefined;
    state.remoteNodeInfo = undefined;
    state.paymentRequest = undefined;
    state.extraData = undefined;
  }),

  queryRoutesForFeeEstimate: thunk(async (_, payload, { injections }) => {
    const queryRoutes = injections.lndMobile.index.queryRoutes;
    return await queryRoutes(payload.pubKey, payload.amount, payload.routeHints);
  }),

  /**
   * @throws
   */
  setPayment: thunk(async (actions, payload, { injections }) => {
    actions.clear();
    const decodePayReq = injections.lndMobile.index.decodePayReq;
    const getNodeInfo = injections.lndMobile.index.getNodeInfo;
    const paymentRequestStr = payload.paymentRequestStr.replace(/^lightning:/i, "");

    try {
      if (!checkBech32(paymentRequestStr, LnBech32Prefix)) {
        throw new Error();
      }
    } catch (e) {
      throw new Error("Code is not a valid Bitcoin Lightning invoice");
    }
    actions.setPaymentRequestStr(paymentRequestStr);

    let paymentRequest;
    try {
      paymentRequest = await decodePayReq(paymentRequestStr);
      actions.setPaymentRequest(paymentRequest);
    } catch (e) {
      throw new Error("Code is not a valid Lightning invoice");
    }

    if (payload.extraData) {
      actions.setExtraData(payload.extraData);
    }

    try {
      const nodeInfo = await getNodeInfo(paymentRequest.destination);
      actions.setRemoteNodeInfo(nodeInfo);
    } catch (e) {}

    return paymentRequest;
  }),

  /**
   * @throws
   */
  sendPayment: thunk(
    async (_, payload, { getState, dispatch, injections, getStoreState, getStoreActions }) => {
      const start = new Date().getTime();

      const sendPaymentV2Sync = injections.lndMobile.index.sendPaymentV2Sync;
      const paymentRequestStr = getState().paymentRequestStr;
      const paymentRequest = getState().paymentRequest;
      const remoteNodeInfo = getState().remoteNodeInfo;
      let outgoingChannelId = undefined;

      if (paymentRequestStr === undefined || paymentRequest === undefined) {
        throw new Error("Payment information missing");
      }

      // If this is a zero sum
      // invoice, hack the value in
      if (!paymentRequest.numSatoshis && payload && payload.amount) {
        paymentRequest.numSatoshis = payload.amount;
        paymentRequest.numMsat = payload.amount.mul(1000);
      }

      if (!!payload && !!payload.outgoingChannelId) {
        outgoingChannelId = payload.outgoingChannelId;
      }

      const extraData: IExtraData = getState().extraData ?? {
        payer: null,
        type: "NORMAL",
        website: null,
        lnurlPayResponse: null,
        lightningAddress: null,
        lud16IdentifierMimeType: null,
        lnurlPayTextPlain: null,
      };

      const name = getStoreState().settings.name;
      const multiPathPaymentsEnabled = getStoreState().settings.multiPathPaymentsEnabled;
      const maxLNFeePercentage = getStoreState().settings.maxLNFeePercentage;
      const getTransactionByPaymentRequest =
        getStoreState().transaction.getTransactionByPaymentRequest;

      // Pre-settlement tx insert
      const preTransaction: ITransaction = getTransactionByPaymentRequest(paymentRequestStr) ?? {
        date: paymentRequest.timestamp,
        description: extraData.lnurlPayTextPlain ?? paymentRequest.description,
        duration: 0,
        expire: paymentRequest.expiry,
        paymentRequest: paymentRequestStr,
        remotePubkey: paymentRequest.destination,
        rHash: paymentRequest.paymentHash,
        status: "OPEN",
        value: paymentRequest.numSatoshis.neg(),
        valueMsat: paymentRequest.numSatoshis.neg().mul(1000),
        amtPaidSat: paymentRequest.numSatoshis.neg(),
        amtPaidMsat: paymentRequest.numSatoshis.neg().mul(1000),
        fee: Long.fromInt(0),
        feeMsat: Long.fromInt(0),
        nodeAliasCached:
          (remoteNodeInfo && remoteNodeInfo.node && remoteNodeInfo.node.alias) || null,
        payer: extraData.payer,
        valueUSD: valueFiat(paymentRequest.numSatoshis, getStoreState().fiat.fiatRates.USD.last),
        valueFiat: valueFiat(paymentRequest.numSatoshis, getStoreState().fiat.currentRate),
        valueFiatCurrency: getStoreState().settings.fiatUnit,
        locationLong: null,
        locationLat: null,
        tlvRecordName: null,
        type: extraData.type,
        website: extraData.website,
        identifiedService: identifyService(
          paymentRequest.destination,
          paymentRequest.description,
          extraData.website,
        ),
        //note: // TODO: Why wasn't this added
        lightningAddress: extraData.lightningAddress ?? null,
        lud16IdentifierMimeType: extraData.lud16IdentifierMimeType ?? null,

        preimage: hexToUint8Array("0"),
        lnurlPayResponse: extraData.lnurlPayResponse,

        hops: [],
      };

      log.d("IPreTransaction", [preTransaction]);
      await dispatch.transaction.syncTransaction(preTransaction);

      let sendPaymentResult: lnrpc.Payment | undefined;
      try {
        sendPaymentResult = await sendPaymentV2Sync(
          paymentRequestStr,
          payload && payload.amount ? Long.fromValue(payload.amount) : undefined,
          paymentRequest.numSatoshis,
          name,
          multiPathPaymentsEnabled,
          maxLNFeePercentage,
          outgoingChannelId,
        );
      } catch (error) {
        await dispatch.transaction.syncTransaction({
          ...preTransaction,
          status: preTransaction.status === "SETTLED" ? preTransaction.status : "CANCELED",
        });
        throw error;
      }

      log.i("status", [sendPaymentResult.status, sendPaymentResult.failureReason]);
      if (sendPaymentResult.status !== lnrpc.Payment.PaymentStatus.SUCCEEDED) {
        await dispatch.transaction.syncTransaction({
          ...preTransaction,
          status: "CANCELED",
        });
        throw new Error(`${translatePaymentFailureReason(sendPaymentResult.failureReason)}`);
      }

      const settlementDuration = new Date().getTime() - start;

      const transaction: ITransaction = {
        ...preTransaction,
        duration: settlementDuration,
        status: "SETTLED",
        fee: sendPaymentResult.fee || Long.fromInt(0),
        feeMsat: sendPaymentResult.feeMsat || Long.fromInt(0),

        preimage: hexToUint8Array(sendPaymentResult.paymentPreimage),

        hops:
          sendPaymentResult.htlcs[0].route?.hops?.map((hop) => ({
            chanId: hop.chanId ?? null,
            chanCapacity: hop.chanCapacity ?? null,
            amtToForward: hop.amtToForward || Long.fromInt(0),
            amtToForwardMsat: hop.amtToForwardMsat || Long.fromInt(0),
            fee: hop.fee || Long.fromInt(0),
            feeMsat: hop.feeMsat || Long.fromInt(0),
            expiry: hop.expiry || null,
            pubKey: hop.pubKey || null,
          })) ?? [],
      };

      log.d("ITransaction", [transaction]);
      await dispatch.transaction.syncTransaction(transaction);

      if (getStoreState().settings.transactionGeolocationEnabled) {
        try {
          log.d("Syncing geolocation for transaction");
          if (PLATFORM === "ios") {
            if (
              (await ReactNativePermissions.check(
                ReactNativePermissions.PERMISSIONS.IOS.LOCATION_WHEN_IN_USE,
              )) === "denied"
            ) {
              log.d("Requesting geolocation permission");
              const r = await ReactNativePermissions.request(
                ReactNativePermissions.PERMISSIONS.IOS.LOCATION_WHEN_IN_USE,
              );
              if (r !== "granted") {
                throw new Error(`Got "${r}" when requesting Geolocation permission`);
              }
            }
          }

          const coords = await getGeolocation();
          transaction.locationLong = coords.longitude;
          transaction.locationLat = coords.latitude;
          await dispatch.transaction.syncTransaction(transaction);
        } catch (error) {
          log.i(`Error getting geolocation for transaction: ${JSON.stringify(error)}`, [error]);
        }
      }

      return sendPaymentResult;
    },
  ),

  setPaymentRequestStr: action((state, payload) => {
    state.paymentRequestStr = payload;
  }),
  setPaymentRequest: action((state, payload) => {
    state.paymentRequest = payload;
  }),
  setRemoteNodeInfo: action((state, payload) => {
    state.remoteNodeInfo = payload;
  }),
  setExtraData: action((state, payload) => {
    state.extraData = payload;
  }),
};

const checkBech32 = (bech32: string, prefix: string): boolean => {
  const decodedBech32 = Bech32.bech32.decode(bech32, 4096);
  if (decodedBech32.prefix.slice(0, prefix.length).toUpperCase() !== prefix.toUpperCase()) {
    return false;
  }
  return true;
};

export const translatePaymentFailureReason = (reason: lnrpc.PaymentFailureReason) => {
  if (reason === lnrpc.PaymentFailureReason.FAILURE_REASON_NONE) {
    throw new Error("Payment failed");
  } else if (reason === lnrpc.PaymentFailureReason.FAILURE_REASON_TIMEOUT) {
    return "Payment timed out";
  } else if (reason === lnrpc.PaymentFailureReason.FAILURE_REASON_NO_ROUTE) {
    return "Could not find route to recipient";
  } else if (reason === lnrpc.PaymentFailureReason.FAILURE_REASON_ERROR) {
    return "The payment failed to proceed";
  } else if (reason === lnrpc.PaymentFailureReason.FAILURE_REASON_INCORRECT_PAYMENT_DETAILS) {
    return "Incorrect payment details provided";
  } else if (reason === lnrpc.PaymentFailureReason.FAILURE_REASON_INSUFFICIENT_BALANCE) {
    return "Insufficient balance";
  }
  return "Unknown error";
};
