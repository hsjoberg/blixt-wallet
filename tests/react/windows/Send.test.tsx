import React from "react";
import { act, render, toJSON, fireEvent, wait, waitForElement } from "@testing-library/react-native";
import Clipboard from "@react-native-community/react-native-clipboard";
import { StoreProvider } from "easy-peasy";
import Long from "long";

import Send from "../../../src/windows/Send";
import SendCamera from "../../../src/windows/Send/SendCamera";
import SendConfirmation from "../../../src/windows/Send/SendConfirmation";
import { createNavigationContainer, setupStore, setDefaultAsyncStorage, mockBlockchainAPI, initCommonStore } from "../../utils";

const AppContainer = createNavigationContainer(Send, "Send");
const AppContainerSendCamera = createNavigationContainer(SendCamera, "SendCamera");
const AppContainerSendConfirmation = createNavigationContainer(SendConfirmation, "SendConfirmation");

it("SendCamera renders correctly", async () => {
  await setDefaultAsyncStorage();

  const store = setupStore();
  await store.getActions().initializeApp();
  await store.getActions().lightning.initialize();
  store.getActions().channel.setBalance(Long.fromNumber(123));

  const { container, unmount } = render(
    <StoreProvider store={store}>
      {AppContainerSendCamera}
    </StoreProvider>
  );
  expect(toJSON(container)).toMatchSnapshot();

  unmount();
});

it("It is possible to paste invoice from clipboard and pay it", async () => {
  await setDefaultAsyncStorage();
  fetch.once(mockBlockchainAPI());
  const store = await initCommonStore(true);

  await store.getActions().initializeApp();
  await store.getActions().lightning.initialize();
  store.getActions().channel.setBalance(Long.fromNumber(10000));
  const invoice = "lntb12u1pww4ckdpp5xck8m9yerr9hqufyd6p0pp0pwjv5nqn6guwr9qf4l66wrqv3h2ssdp2xys9xct5da3kx6twv9kk7m3qg3hkccm9ypxxzar5v5cqp5ynhgvxfnkwxx75pcxcq2gye7m5dj26hjglqmhkz8rljhg3eg4hfyg38gnsynty3pdatjg9wpa7pe7g794y0hxk2gqd0hzg2hn5hlulqqen6cr5";
  Clipboard.setString(invoice);

  const { queryByTestId, unmount } = render(
    <StoreProvider store={store}>
      {AppContainer}
    </StoreProvider>
  );

  const pasteClipboardIcon = queryByTestId("paste-clipboard");
  expect(pasteClipboardIcon).not.toBeNull();

  await act(async () => await fireEvent.press(pasteClipboardIcon!));

  const payInvoiceButton = await waitForElement(() => queryByTestId("pay-invoice"));
  await act(async () => await fireEvent.press(payInvoiceButton!));

  expect(store.getState().transaction.transactions).toHaveLength(1);

  unmount();
});
