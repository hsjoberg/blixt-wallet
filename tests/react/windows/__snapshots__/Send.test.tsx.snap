// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendCamera renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "zIndex": 1,
          },
          false,
        ]
      }
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        onLayout={[Function]}
        pointerEvents="box-none"
        style={null}
      >
        <View
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": 0,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(255, 255, 255)",
                "borderBottomColor": "rgb(216, 216, 216)",
                "flex": 1,
                "shadowColor": "rgb(216, 216, 216)",
                "shadowOffset": Object {
                  "height": 0.5,
                  "width": 0,
                },
                "shadowOpacity": 0.85,
                "shadowRadius": 0,
              }
            }
          />
        </View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "height": 64,
              "maxHeight": undefined,
              "minHeight": undefined,
              "opacity": undefined,
              "transform": undefined,
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "height": 20,
              }
            }
          />
          <View
            pointerEvents="box-none"
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "marginHorizontal": 16,
                  "opacity": 1,
                }
              }
            >
              <Text
                accessibilityRole="header"
                numberOfLines={1}
                onLayout={[Function]}
                style={
                  Object {
                    "color": "rgb(28, 28, 30)",
                    "fontSize": 17,
                    "fontWeight": "600",
                  }
                }
              >
                SendCamera
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "opacity": 1,
            }
          }
        />
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onGestureEnd={[Function]}
          onOpen={[Function]}
          onTransitionStart={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              collapsable={false}
              forwardedRef={[Function]}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              style={
                Object {
                  "flex": 1,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "rgb(242, 242, 242)",
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": -1,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 5,
                    "top": 0,
                    "width": 3,
                  }
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "backgroundColor": "rgb(242, 242, 242)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "column-reverse",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "bottom": 0,
                          "left": 0,
                          "position": "absolute",
                          "right": 0,
                          "top": 0,
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "alignItems": "center",
                              "bottom": 0,
                              "justifyContent": "center",
                              "left": 0,
                              "position": "absolute",
                              "right": 0,
                              "top": 0,
                            },
                          ]
                        }
                      >
                        <View
                          style={
                            Array [
                              Object {
                                "alignItems": "center",
                                "justifyContent": "center",
                              },
                              Object {
                                "height": 275,
                                "width": 275,
                              },
                            ]
                          }
                        >
                          <View
                            style={
                              Array [
                                Object {
                                  "borderColor": "#bc6610",
                                  "height": 20,
                                  "width": 20,
                                },
                                Object {
                                  "left": 0,
                                  "position": "absolute",
                                  "top": 0,
                                },
                                Object {
                                  "borderLeftWidth": 4,
                                  "borderTopWidth": 4,
                                },
                              ]
                            }
                          />
                          <View
                            style={
                              Array [
                                Object {
                                  "borderColor": "#bc6610",
                                  "height": 20,
                                  "width": 20,
                                },
                                Object {
                                  "position": "absolute",
                                  "right": 0,
                                  "top": 0,
                                },
                                Object {
                                  "borderRightWidth": 4,
                                  "borderTopWidth": 4,
                                },
                              ]
                            }
                          />
                          <View
                            style={
                              Array [
                                Object {
                                  "borderColor": "#bc6610",
                                  "height": 20,
                                  "width": 20,
                                },
                                Object {
                                  "bottom": 0,
                                  "left": 0,
                                  "position": "absolute",
                                },
                                Object {
                                  "borderBottomWidth": 4,
                                  "borderLeftWidth": 4,
                                },
                              ]
                            }
                          />
                          <View
                            style={
                              Array [
                                Object {
                                  "borderColor": "#bc6610",
                                  "height": 20,
                                  "width": 20,
                                },
                                Object {
                                  "bottom": 0,
                                  "position": "absolute",
                                  "right": 0,
                                },
                                Object {
                                  "borderBottomWidth": 4,
                                  "borderRightWidth": 4,
                                },
                              ]
                            }
                          />
                        </View>
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "height": "100%",
                              "justifyContent": "space-around",
                              "left": 0,
                              "position": "absolute",
                              "top": 0,
                              "width": "100%",
                            }
                          }
                        >
                          <View
                            style={
                              Array [
                                Object {
                                  "width": "100%",
                                },
                                Object {
                                  "backgroundColor": "rgba(0, 0, 0, 0.6)",
                                  "flex": 1,
                                },
                              ]
                            }
                          />
                          <View
                            onLayout={[Function]}
                            style={
                              Array [
                                Object {
                                  "height": 275,
                                },
                                Object {
                                  "display": "flex",
                                  "flexDirection": "row",
                                },
                              ]
                            }
                          >
                            <View
                              style={
                                Array [
                                  Object {
                                    "backgroundColor": "rgba(0, 0, 0, 0.6)",
                                    "flex": 1,
                                  },
                                ]
                              }
                            />
                            <View
                              style={
                                Array [
                                  Object {
                                    "backgroundColor": "transparent",
                                  },
                                  Object {
                                    "height": 275,
                                    "width": 275,
                                  },
                                ]
                              }
                            />
                            <View
                              style={
                                Array [
                                  Object {
                                    "backgroundColor": "rgba(0, 0, 0, 0.6)",
                                    "flex": 1,
                                  },
                                ]
                              }
                            />
                          </View>
                          <View
                            style={
                              Array [
                                Object {
                                  "width": "100%",
                                },
                                Object {
                                  "backgroundColor": "rgba(0, 0, 0, 0.6)",
                                  "flex": 1,
                                },
                              ]
                            }
                          />
                        </View>
                      </View>
                      <Text
                        allowFontScaling={false}
                        onPress={[Function]}
                        style={
                          Array [
                            Object {
                              "color": undefined,
                              "fontSize": 12,
                            },
                            Array [
                              Object {
                                "color": "#000",
                                "fontSize": 30,
                              },
                              Object {
                                "color": "#000",
                                "fontSize": 30,
                              },
                              Object {
                                "bottom": 7,
                                "color": "#e5eaea",
                                "fontSize": 26,
                                "left": 8,
                                "padding": 4,
                                "position": "absolute",
                              },
                            ],
                            Object {
                              "fontFamily": "Ionicons",
                              "fontStyle": "normal",
                              "fontWeight": "normal",
                            },
                            Object {},
                          ]
                        }
                        type="Ionicons"
                      >
                        
                      </Text>
                      <Text
                        allowFontScaling={false}
                        onPress={[Function]}
                        style={
                          Array [
                            Object {
                              "color": undefined,
                              "fontSize": 12,
                            },
                            Array [
                              Object {
                                "color": "#000",
                                "fontSize": 30,
                              },
                              Object {
                                "color": "#000",
                                "fontSize": 30,
                              },
                              Object {
                                "bottom": 8,
                                "color": "#e5eaea",
                                "fontSize": 26,
                                "padding": 4,
                                "position": "absolute",
                                "right": 64,
                              },
                            ],
                            Object {
                              "fontFamily": "Material Design Icons",
                              "fontStyle": "normal",
                              "fontWeight": "normal",
                            },
                            Object {},
                          ]
                        }
                        type="MaterialCommunityIcons"
                      >
                        󰆷
                      </Text>
                      <Text
                        allowFontScaling={false}
                        onPress={[Function]}
                        style={
                          Array [
                            Object {
                              "color": undefined,
                              "fontSize": 12,
                            },
                            Array [
                              Object {
                                "color": "#000",
                                "fontSize": 30,
                              },
                              Object {
                                "color": "#000",
                                "fontSize": 30,
                              },
                              Object {
                                "bottom": 8,
                                "color": "#e5eaea",
                                "fontSize": 26,
                                "padding": 4,
                                "position": "absolute",
                                "right": 8,
                              },
                            ],
                            Object {
                              "fontFamily": "FontAwesome",
                              "fontStyle": "normal",
                              "fontWeight": "normal",
                            },
                            Object {},
                          ]
                        }
                        testID="paste-clipboard"
                        type="FontAwesome"
                      >
                        
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
